content/blog/2019-10-18-double-checked-locking/index.md

```diff
@@ -0,0 +1,491 @@
+                if (h == null)
+                    synchronized(this) {
+                        ptr = allocate();
+                        h = ptr;
+                        helper = h;
+                        ptr.field1 = initField1();
+                        ptr.field2 = initField2();
+                    }                       
+            }
+        }
+        return helper;
+    }
+}
+```
+
+## Working Solutions
+
+### Explicit Memory Barrier
+
+The previous fix with two synchronized sections does not work 
+because releasing lock is an implicit "one-way" memory barrier.
+It is possible to make the double-checked locking actually work with an
+explicit memory barrier.
+For example, in C++11 we can safely implement double-checked locking with 
+`std::atomic` and `std::atomic_thread_fence`.
+
+```C++
+class Foo {
+    private:
+        std::atomic <Foo*> helper;
+    public:
+        Foo* get_helper() {
+            Foo* h = helper.load(std::memory_order_relaxed);
+            std::atomic_thread_fence(std::memory_order_acquire);        //memory barrier
+            if (h == nullptr) {
+                std::lock_guard<std::mutex> lock(m_init);
+                h = helper.load(std::memory_order_relaxed);
+                if (h == nullptr) {
+                    h = new Helper;
+                    std::atomic_thread_fence(std::memory_order_release);//memory barrier
+                    helper.store(h, std::memory_order_relaxed);
+                }
+            }
+            return tmp;
+        }
+};
+```
+
+* `std::atomic_thread_fence(std::memory_order_acquire)` guarantees that 
+read/write operations after a memory barrier cannot be reordered with 
+read operations before the memory barrier.
+* `std::atomic_thread_fence(std::memory_order_release)` guarantees that 
+read/write operations before a memory barrier cannot be reordered with
+write operations after the memory barrier.
+
+The memory barriers guarantee that 
+1. `h` loads the value from `helper` before starting object initialization.
+2. object initialization finishes before storing the value to `helper`. 
+
+### Atomic Operation
+
+An atomic operation will either happen completely, or it does not happen at all.
+This is no intermediate state, so that the side effects of an atomic operation
+will not be visible until the operation is complete.
+
+In previous analysis, we have seen that `h = new Helper()` can be interleaved
+because it is not an atomic operations.
+If this operation is atomic, the double-checked locking will work.
+
+#### 32-bit Primitive Variables
+
+Read and write operations of most primitive variables 
+(except `long` and `double` since they are 64-bit)are atomic.
+If the initialized value is a 32-bit primitive variable,
+   assignment to the variable will only happen once the data is available.
+Since the write operation is atomic, 
+other threads will either see a ready-to-use value or 0.
+There is no middle state of "initializing".
```

Comment:
Ah I see the point. Since it is not `private volatile int magicNumber`, but it somehow behaves like a volatile variable, so it needs some specific memory model support from Java.